File: problems/Trapping Rain Water/trap.cs
public class Solution {
    public int Trap(int[] height) {
        int n = height.Length;
        int[] leftToRight = new int[n];
        int[] rightToLeft = new int[n];
        int result = 0;

        for (int idx = 0; n > idx; ++idx) {
            if (0 == idx || leftToRight[idx - 1] < height[idx]) {
                leftToRight[idx] = height[idx];
            } else {
                leftToRight[idx] = leftToRight[idx - 1];
            }

            if (0 == idx || rightToLeft[n - idx] < height[n - idx - 1]) {
                rightToLeft[n - idx - 1] = height[n - idx - 1];
            } else {
                rightToLeft[n - idx - 1] = rightToLeft[n - idx];
            }
        }

        for (int idx = 0; n > idx; ++idx) {
            result += Math.Min(leftToRight[idx], rightToLeft[idx]) - height[idx];
        }

        return result;
    }
}

File: problems/Trapping Rain Water/trap.java
class Solution {
    public int trap(int[] height) {
        int result = 0;
        int n = height.length;

        if (0 == n) {
            return result;
        }

        int[] leftToRight = new int[n];
        int[] rightToLeft = new int[n];

        leftToRight[0] = height[0];
        rightToLeft[n - 1] = height[n - 1];

        for (int idx = 1; n > idx; ++idx) {
            leftToRight[idx] = Math.max(leftToRight[idx - 1], height[idx]);
            rightToLeft[n - 1 - idx] = Math.max(rightToLeft[n - idx], height[n - 1 - idx]);
        }

//         System.out.println(Arrays.stream(leftToRight)
//                  .mapToObj(String::valueOf)
//                  .collect(Collectors.joining(", ")));

//         System.out.println(Arrays.stream(rightToLeft)
//                  .mapToObj(String::valueOf)
//                  .collect(Collectors.joining(", ")));

        for (int idx = 0; n > idx; ++idx) {
            result += Math.min(leftToRight[idx], rightToLeft[idx]) - height[idx];
        }

        return result;
    }
}
